File: app/src/main/java/com/salman/bitclock/di/AppModule.java
package com.salman.bitclock.di;

import android.app.Application;
import android.content.Context;

import com.salman.bitclock.data.repository.AlarmRepository;
import com.salman.bitclock.data.AppStateManager;
import com.salman.bitclock.data.database.AlarmDao;
import com.salman.bitclock.data.database.AppDatabase;
import com.salman.bitclock.data.database.TimerDao;
import com.salman.bitclock.utils.AlarmScheduler;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.android.qualifiers.ApplicationContext;
import dagger.hilt.components.SingletonComponent;
import dagger.hilt.EntryPoint;

/**
 * Dependency Injection module using Hilt.
 * Provides singleton instances for the app's core components.
 */
@Module
@InstallIn(SingletonComponent.class)
public class AppModule {

    /**
     * EntryPoint for accessing AlarmRepository in non-Hilt classes (e.g., BroadcastReceivers).
     */
    @EntryPoint
    @InstallIn(SingletonComponent.class)
    public interface AlarmRepositoryEntryPoint {
        AlarmRepository alarmRepository();
    }

    @Provides
    @Singleton
    public AppDatabase provideDatabase(Application application) {
        return AppDatabase.getDatabase(application);
    }

    @Provides
    @Singleton
    public AlarmDao provideAlarmDao(AppDatabase appDatabase) {
        return appDatabase.alarmDao();
    }

    @Provides
    @Singleton
    public TimerDao provideTimerDao(AppDatabase appDatabase) {
        return appDatabase.timerDao();
    }

    @Provides
    @Singleton
    public AppStateManager provideAppStateManager() {
        return new AppStateManager();
    }

    @Provides
    @Singleton
    public AlarmScheduler provideAlarmScheduler(@ApplicationContext Context context) {
        return new AlarmScheduler(context);
    }
}

File: app/src/main/kotlin/com/salman/bitclock/di/AppModule.kt
package com.salman.bitclock.di

import android.app.Application
import android.content.Context
import com.salman.bitclock.data.database.AlarmDao
import com.salman.bitclock.data.database.AppDatabase
import com.salman.bitclock.data.database.TimerDao
import com.salman.bitclock.data.database.WorldClockDao
import com.salman.bitclock.data.repository.AlarmRepository
import com.salman.bitclock.utils.AlarmScheduler
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import dagger.hilt.EntryPoint
import javax.inject.Singleton

@Module
@InstallIn(SingletonComponent::class)
object AppModule {

    @EntryPoint
    @InstallIn(SingletonComponent::class)
    interface AlarmRepositoryEntryPoint {
        fun alarmRepository(): AlarmRepository
    }

    @Provides
    @Singleton
    fun provideDatabase(application: Application): AppDatabase {
        return AppDatabase.getDatabase(application)
    }

    @Provides
    @Singleton
    fun provideAlarmDao(database: AppDatabase): AlarmDao = database.alarmDao()

    @Provides
    @Singleton
    fun provideTimerDao(database: AppDatabase): TimerDao = database.timerDao()

    @Provides
    @Singleton
    fun provideWorldClockDao(database: AppDatabase): WorldClockDao = database.worldClockDao()

    @Provides
    @Singleton
    fun provideSleepDao(database: AppDatabase): com.salman.bitclock.data.database.SleepDao = database.sleepDao()

    @Provides
    @Singleton
    fun provideHabitDao(database: AppDatabase): com.salman.bitclock.data.database.HabitDao = database.habitDao()

    @Provides
    @Singleton
    fun provideProfileDao(database: AppDatabase): com.salman.bitclock.data.database.ProfileDao = database.profileDao()

    @Provides
    @Singleton
    fun provideAuditLogDao(database: AppDatabase): com.salman.bitclock.data.database.AuditLogDao = database.auditLogDao()

    @Provides
    @Singleton
    fun provideAlarmScheduler(@ApplicationContext context: Context): AlarmScheduler {
        return AlarmScheduler(context)
    }
}
